## 1. Get data

Using curl (or Invoke-RestMethod when on powershell), call the Google Books API and answer the following questions by examing the json responses.

	- How many pages does book with ISBN 9780884271956 have?
	- When was "The Phoenix Project" by Gene Kim published?

In [1]:
#I got this code from :
#https://gist.github.com/IT102Gists/aa31be3ac569878cc7df3a8b37ae25a4

# Python's built-in module for encoding and decoding JSON data
import json
# Python's built-in module for opening and reading URLs
from urllib.request import urlopen

# sample ISBN for testing: 1593276036
# "https://www.googleapis.com/books/v1/volumes/s1gVAAAAYAAJ"
    
while True:

    # create getting started variables
    api = "https://www.googleapis.com/books/v1/volumes?q=isbn:"
    isbn = input("Enter 10 or 13 digit ISBN: ").strip()

    # send a request and get a JSON response
    resp = urlopen(api + isbn)
    # parse JSON into Python as a dictionary
    book_data = json.load(resp)

    # create additional variables for easy querying
    volume_info = book_data["items"][0]["volumeInfo"]
    author = volume_info["authors"]
    # practice with conditional expressions
    prettify_author = author if len(author) > 1 else author[0]

    # display title, author, page count, publication date
    # fstrings require Python 3.6 or higher
    # \n adds a new line for easier reading
    print(f"\nTitle: {volume_info['title']}")
    print(f"Author: {prettify_author}")
    print(f"Page Count: {volume_info['pageCount']}")
    print(f"Publication Date: {volume_info['publishedDate']}")
    print("\n***\n")

    # ask user if they would like to enter another isbn
    user_update = input("Would you like to enter another ISBN? y or n ").lower().strip()

    if user_update != "y":
        print("May the Zen of Python be with you. Have a nice day!")
        break # as the name suggests, the break statement breaks out of the while loop

Enter 10 or 13 digit ISBN: 1593276036

Title: Python Crash Course
Author: Eric Matthes
Page Count: 560
Publication Date: 2015-11-20

***

Would you like to enter another ISBN? y or n y
Enter 10 or 13 digit ISBN: 9780884271956

Title: The Goal
Author: ['Eliyahu M. Goldratt', 'Jeff Cox', 'David Whitford']
Page Count: 408
Publication Date: 2012-06-01

***

Would you like to enter another ISBN? y or n n
May the Zen of Python be with you. Have a nice day!


## Challenge Answer


### The book with ISBN 9780884271956 has 408 pages!


In [2]:
# the quote module is needed to prevent the error "URL can't contain control characters"
from urllib.parse import quote

boolean = True
while boolean:

    # create getting started variables
    ### TRY TO FIX THIS PART. WHY DOESNT inttitle:?inauthor work in the api??
    api = "https://www.googleapis.com/books/v1/volumes?q=intitle:?inauthor:"
    intitle = input("Enter the title name: ").strip()
    inauthor = input("Enter the author name: ").strip()
    boolean = False
    # quote(param) makes sure that the input works in the api query
    resp = urlopen( api + quote(intitle)  
                  # + quote(inauthor)
                  )
    
    book_data = json.load(resp)

    volume_info = book_data["items"][0]["volumeInfo"]
    publish_info = volume_info["publishedDate"]
    print(f"\n The Published date for {intitle} by {inauthor} is: {publish_info}")


Enter the title name: The Phoenix Project
Enter the author name: Gene Kim

 The Published date for The Phoenix Project by Gene Kim is: 2016-04-04


In [3]:
# This is some code to play around with the q={params}
## Why do I can different results if I change the api params such as & as & or +. 
### Advice from Miguel is to ask Frank Boogaard
api = ( "https://www.googleapis.com/books/v1/volumes?q=intitle:The+Phoenix+Project&inauthor:Gene+Kim")

resp = urlopen ( api )

data = json.load(resp)
publish_info = data["items"][0]["volumeInfo"]["publishedDate"]
published_info_two = data["items"][1]["volumeInfo"]["publishedDate"]

print(publish_info)
print(published_info_two)
print(f"\nThe Published date for The Phoenix Project by Gene Kim is: {publish_info}")

2016-04-04
2014-10-14

The Published date for The Phoenix Project by Gene Kim is: 2016-04-04


## Challenge Answer Two
### The Publish date for The Phoenix Project by Gene Kim is 2016-04-04 

# Scripting API calls
## 2. Get title by ISBN

Write a script in Python that shows the title of a book for a given ISBN number.

In [5]:
    # create getting started variables
    api = "https://www.googleapis.com/books/v1/volumes?q=isbn:"
    isbn = input("Enter 10 or 13 digit ISBN: ").strip()

    # send a request and get a JSON response
    resp = urlopen(api + isbn)
    # parse JSON into Python as a dictionary
    book_data = json.load(resp)
    # create additional variables for easy querying
    volume_info = book_data["items"][0]["volumeInfo"]
    title = volume_info["title"]

    print(f"\nThe title associated to the isbn number {isbn} is {title}")

Enter 10 or 13 digit ISBN: 9780884271956

The title associated to the isbn number 9780884271956 is The Goal


## Challenge Answer Three 

# Loop through data
## 3. Get titles by ISBN list

Improve script to show titles of books for the following ISBN numbers:  
* `["9780141036144","9780099518471", "9780099560432", "9781784971571"]`

In [6]:
isbn_list = ["9780141036144","9780099518471", "9780099560432", "9781784971571"]

api = "https://www.googleapis.com/books/v1/volumes?q=isbn:"

for i in isbn_list: 
    # send a request and get a JSON response
    resp = urlopen(api + i)
    # parse JSON into Python as a dictionary
    book_data = json.load(resp)
    title = book_data["items"][0]["volumeInfo"]["title"]
    print(f"\n The title of isbn number {i} is {title}")



 The title of isbn number 9780141036144 is Nineteen Eighty-four

 The title of isbn number 9780099518471 is Brave New World

 The title of isbn number 9780099560432 is Ready Player One

 The title of isbn number 9781784971571 is The Three-Body Problem


## 4. Show specific book data

Using the Google Books API, write a script that given an author, does the following:

* Show the amount of results that match the query
* Print ID, Title and ISBN number to screen

In [7]:
from urllib.parse import quote

boolean = True
while boolean:

    # create getting started variables
    api = "https://www.googleapis.com/books/v1/volumes?q=inauthor:"
#     firstname = input("Enter the authors' first name: ").strip()
#     lastname = input("Enter the authors' last name: ").strip()
#     resp = urlopen( api + firstname + ' ' +lastname )
    
    inauthor = input("Enter the authors' name: ").strip()
    boolean = False
    resp = urlopen( api + quote(inauthor) )
    
    book_data = json.load(resp)
   
    # Query through the JSON format
    total_items = book_data["totalItems"]
    volume_info = book_data["items"][0]["volumeInfo"]
    id_info = book_data["items"][0]["id"]
    title_info = volume_info["title"]
    isbn_info = volume_info["industryIdentifiers"]
    
    print(f"\n The total items that matched the api request is: {total_items}")
    print(f"\n ID: {id_info} \n Title: {title_info} \n ISBN: {isbn_info}")

Enter the authors' name: Gene Kim

 The total items that matched the api request is: 507

 ID: ui8hDgAAQBAJ 
 Title: The DevOps Handbook 
 ISBN: [{'type': 'ISBN_13', 'identifier': '9781942788072'}, {'type': 'ISBN_10', 'identifier': '194278807X'}]


# Bonus: Create yaml
#### 5. Create a yaml file based on the book list in question 3, that contains ID, Title, ISBN and amount of page (if present, otherwise set it to 'N/A').
* Link to solution on stackoverflow https://stackoverflow.com/questions/38781929/how-to-convert-json-to-yaml-in-javascript

In [9]:
import sys
import ruamel.yaml
from urllib.parse import quote


def non_quoted_key(self, data):
   if isinstance(data, ruamel.yaml.compat.string_types):
       data = ruamel.yaml.scalarstring.PlainScalarString(data)
   return self.represent_data(data)

boolean = True
while boolean:

    # create getting started variables
    api = "https://www.googleapis.com/books/v1/volumes?q=inauthor:"
    
    inauthor = input("Enter the authors' name: ").strip()
    boolean = False
    resp = urlopen( api + quote(inauthor) )
    
    book_data = json.load(resp)

    yaml = ruamel.yaml.YAML()
    yaml.default_flow_style = False
    yaml.default_style = '"'
    yaml.Representer.represent_key = non_quoted_key
    yaml.dump(book_data, sys.stdout)

Enter the authors' name: Gene Kim
kind: "books#volumes"
totalItems: !!int "507"
items:
- kind: "books#volume"
  id: "ui8hDgAAQBAJ"
  etag: "YLkE3HRVVH4"
  selfLink: "https://www.googleapis.com/books/v1/volumes/ui8hDgAAQBAJ"
  volumeInfo:
    title: "The DevOps Handbook"
    subtitle: "How to Create World-Class Agility, Reliability, and Security in Technology\
      \ Organizations"
    authors:
    - "Gene Kim"
    - "Jez Humble"
    - "Patrick Debois"
    - "John Willis"
    publisher: "IT Revolution"
    publishedDate: "2016-10-06"
    description: "Increase profitability, elevate work culture, and exceed productivity\
      \ goals through DevOps practices. More than ever, the effective management of\
      \ technology is critical for business competitiveness. For decades, technology\
      \ leaders have struggled to balance agility, reliability, and security. The\
      \ consequences of failure have never been greater―whether it's the healthcare.gov\
      \ debacle, cardholder 